pkg/controller/endpointslice/reconciler.go

```diff
@@ -0,0 +1,288 @@
+		err := r.client.DiscoveryV1alpha1().EndpointSlices(r.service.Namespace).Delete(sliceName, &metav1.DeleteOptions{})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error deleting %s EndpointSlice for Service %s/%s: %v", sliceName, r.service.Namespace, r.service.Name, err))
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (r *Reconciler) reconcileByPortHash(
+	existingSlices []discovery.EndpointSlice,
+	desiredEndpoints []discovery.Endpoint,
+	endpointMeta EndpointMeta,
+) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice, sets.String) {
+	// build data structures for desired state
+	endpointsByHash := map[string]discovery.Endpoint{}
+	desiredEndpointHashes := sets.String{}
+	for _, desiredEndpoint := range desiredEndpoints {
+		eHash := endpointToHash(desiredEndpoint)
+		desiredEndpointHashes.Insert(eHash)
+		endpointsByHash[eHash] = desiredEndpoint
+	}
+
+	// build data structures for existing state
+	existingEndpointHashes := sets.String{}
+	endpointHashesBySliceName := map[string]sets.String{}
+	for _, existingSlice := range existingSlices {
+		endpointHashesBySliceName[existingSlice.Name] = sets.String{}
+		for _, endpoint := range existingSlice.Endpoints {
+			eHash := endpointToHash(endpoint)
+			endpointsByHash[eHash] = endpoint
+			existingEndpointHashes.Insert(eHash)
+			endpointHashesBySliceName[existingSlice.Name].Insert(eHash)
+		}
+	}
+
+	epHashesToAdd := desiredEndpointHashes.Difference(existingEndpointHashes)
+	epHashesToRemove := existingEndpointHashes.Difference(desiredEndpointHashes)
+
+	sliceNamesToUpdate := sets.String{}
+	sliceNamesToDelete := sets.String{}
+
+	for name := range endpointHashesBySliceName {
+		// remove any endpoints that are no longer desired
+		if epHashesToRemove.Len() > 0 {
+			origLength := endpointHashesBySliceName[name].Len()
+			endpointHashesBySliceName[name] = endpointHashesBySliceName[name].Difference(epHashesToRemove)
+
+			// if slice has changed, add as many desired endpoints as it can fit
+			if endpointHashesBySliceName[name].Len() != origLength {
+				diff := int(r.maxEndpointsPerSlice) - endpointHashesBySliceName[name].Len()
+				for epHashesToAdd.Len() > 0 && diff > 0 {
+					epHash, _ := epHashesToAdd.PopAny()
+					endpointHashesBySliceName[name].Insert(epHash)
+					diff--
+				}
+
+				// if slice has endpoints left, update slice, otherwise mark it for removal
+				if endpointHashesBySliceName[name].Len() > 0 {
+					sliceNamesToUpdate.Insert(name)
+				} else {
+					sliceNamesToDelete.Insert(name)
+				}
+			}
+		}
+	}
+
+	// We want to minimize unnecessary writes instead of trying to top up every
+	// existing slice. If an existing slice has changed, we top it up as much as
+	// possible (see above). If there are still desired endpoints leftover at
+	// this point, we try to fit the remainder (% maxEndpointsPerSlice) into an
+	// existing slice that will be closest to full.
+	leftover := epHashesToAdd.Len() % int(r.maxEndpointsPerSlice)
+
+	if leftover > 0 {
```

Comment:
Please correct me if I am wrong, I think the sequence of steps may not be accurate here. 

Let us say: 
Existing EndpointSlice has 80 endpoints. And leftover = 101, 
Then it will result in creating 2 endpoint slices with 100 endpoints and 1 endpoint.

Reply:
probably this is better:

for leftover >0 {
     if leftover > 100:  create new endpointSlice
     if leftover < 100:
                slice, ok :=  findSliceCandidate
                if !ok {
                        create new EndpointSlice
                 }

}

Reply:
This was a confusing part, leftover here referred to `(endpointsToAdd.Len() % maxEndpointsPerSlice)` and creation happened later, cleaning this up to be less confusing.